(3/5) Multiple Issues: Close gaps after mouse button release

andreasfelix opened the pull request:
**(3/5) Split up #177 in different PRs:**

Needs to be merged after #181

**Changes in this PR:**
- close gaps after mouse button release for up to 3 monitors (#35.1)

![close-gaps](https://user-images.githubusercontent.com/24651767/63471216-d6297a80-c46e-11e9-9abe-e0924d66df2f.gif)

jeremypw commented on src/Widgets/DisplaysOverlay.vala:
```diff
@@ -275,6 +281,47 @@ public class Display.DisplaysOverlay : Gtk.Overlay {
+            Gdk.Rectangle other_rect = { other_x, other_y, other_width, other_height };
+            Gdk.Rectangle intersection;
+            var is_connected = rect.intersect (other_rect, out intersection);
+            var is_diagonal = intersection.height == 1 && intersection.width == 1;
```
Should be `var is_diagonal = is_connected && (intersection.height <= 50 && intersection.width <= 50);` to regard "forbidden" formation as disconnected. Probably best to make the `50` a const variable for clarity.

andreasfelix replied:
- made 50 a const
- for first point see comment

andreasfelix commented on src/Widgets/DisplaysOverlay.vala:
```diff
@@ -384,116 +431,62 @@ public class Display.DisplaysOverlay : Gtk.Overlay {
+                    test_distance_x += diff_x > 0 ? 50 : -50;
+                } else {
+                    test_distance_y += diff_y > 0 ? 50 : -50;
+                }
```
I renamed diff_x/y to distance_widget_anchor_x/y to make variable names more explanatory. Do you have some other suggestions for better variable names?

I think the difficulty to name geometry stuff is that there are 100 different equivalent ways to look at same problem. So a formula which seem intuitive someone might appear counter-intuitive to someone else...

jeremypw replied:
Then comments to indicate the intention of each formula are needed.

andreasfelix commented:
@jeremypw I intentional used the 50px offset only in the snapping function. My idea was that it seems like a reasonable default and makes it clear to the user that diagonal monitors are not allowed (1px would be hard to notice). But if a user wants to place their monitor within the 1-49 px margin (which technically allowed), I see no reason to explicitly forbid this.

But if you have a strong opinion that the 50px offset should be mandatory, I will change it. I just wanted to let you know that I already took it into consideration...

jeremypw commented:
@andreasfelix Regarding diagonal offsets, I am happy to go with your way for now.  No doubt an issue will be raised if there is any serious objection. 
With regards to comments, all I can say is that I found it hard to understand the code without adding comments of my own to remind me what each variable represented.  Variables `distance_x` and `test_distance_x` etc are hard to understand without a comment. A picture would be even better! Variable names of reasonable length that are fully self documenting does not come to mind I am afraid.

jeremypw commented on src/Widgets/DisplaysOverlay.vala:
```diff
@@ -384,116 +433,62 @@ public class Display.DisplaysOverlay : Gtk.Overlay {
+            var test_distance_x = distance_right > -distance_left ? distance_right : distance_left;
+            var test_distance_y = distance_bottom > -distance_top ? distance_bottom : distance_top;
+
+            if (distance_left > 0 && distance_right < 0) {
```
Comment that this meams widget is between left and right edges of anchor so no horizontal movement needed.

jeremypw commented on src/Widgets/DisplaysOverlay.vala:
```diff
@@ -384,116 +433,62 @@ public class Display.DisplaysOverlay : Gtk.Overlay {
+
+            if (distance_left > 0 && distance_right < 0) {
+                test_distance_x = 0;
+            } else if (distance_top > 0 && distance_bottom < 0) {
```
Comment that this meams widget is between top and bottom edges of anchor so no vertical movement needed.

jeremypw commented on src/Widgets/DisplaysOverlay.vala:
```diff
@@ -384,116 +433,62 @@ public class Display.DisplaysOverlay : Gtk.Overlay {
+            var distance_right  = distance_widget_anchor_x - widget_width;
+            var distance_top    = distance_widget_anchor_y + anchor_height;
+            var distance_bottom = distance_widget_anchor_y - widget_height;
+            var test_distance_x = distance_right > -distance_left ? distance_right : distance_left;
```
Isn't this the same as 
`var test_distance_x = distance_right.abs () < distance_left.abs () ? distance_right : distance_left;`

If so, it is more obvious that `test_distance_x` is in fact the smallest horizontal gap and could perhaps be renamed as such.

andreasfelix replied:
- Yes, it is the same
- I renamed `test_distance_x`

andreasfelix commented:
@jeremypw, I have added the comments where it is checked if the widget is within the edges of the anchor. I have also renamed the `distance_x/y` and `test_distance_x/y` variables. I think their purpose should be more clear now.

If there are any other parts which are difficult to understand, I think it would be the best (if it is okay for you) if you would comment them. For me it is not obvious what could be unclear and I am afraid I would add to many unnecessary comments.

jeremypw commented on src/Widgets/DisplaysOverlay.vala:
```diff
@@ -275,6 +282,48 @@ public class Display.DisplaysOverlay : Gtk.Overlay {
     }
 
+    private void close_gaps () {
+        var display_widgets = new List<DisplayWidget>();
```
Lint: missing space

jeremypw commented on src/Widgets/DisplaysOverlay.vala:
```diff
@@ -384,116 +433,65 @@ public class Display.DisplaysOverlay : Gtk.Overlay {
+            var distance_origin_y = anchor_y - widget_y;
+            var distance_left   = distance_origin_x + anchor_width;
+            var distance_right  = distance_origin_x - widget_width;
+            var distance_top    = distance_origin_y + anchor_height;
```
Lint: multiple spaces instead of single space in above 3 lines.

jeremypw commented on src/Widgets/DisplaysOverlay.vala:
```diff
@@ -384,116 +433,65 @@ public class Display.DisplaysOverlay : Gtk.Overlay {
-            if (snap_x & move) {
-                debug ("moving child %d on X\n", shortest_x);
-                if (shortest_x < SNAP_LIMIT) ((DisplayWidget) child).set_geometry (child_x - shortest_x, child_y , child_width, child_height);
+            var shortest_distance_candidate = distance_widget_anchor_x * distance_widget_anchor_x + distance_widget_anchor_y * distance_widget_anchor_y;
```
Lint: Significantly exceeds 120 char line length
